### PR TITLE
Substituer Pandas par SQLite pour le traitement de données

### LINKED ISSUE #30: Créer la base SQLite

### LINKED ISSUE #31: Créer le schema de base de données

### LINKED ISSUE #38: Remplacer Jupyter Notebook par des scripts Python

### LINKED ISSUE #40: Optimiser l'usage mémoire du preprocessing

### LINKED ISSUE #45: Remplacer les opérateurs Elasticsearch par des scripts Python

### LINKED ISSUE #46: Indexer les données par chunk depuis SQLite

### PR DESCRIPTION
**Pull request by HAEKADI**
closes #38 
closes #30
closes #40
closes #31 
closes #45 
closes #46

### CODE REVIEW

File: DAG-insert-elk-sirene.py [hunk 1]
```diff
@@ -102,12 +148,12 @@
...
-    send_email.set_upstream(execute_aio_container)
+    create_sqlite_database << clean_previous_folder << get_colors
+    create_etablissement_table << create_unite_legale_table << create_sqlite_database
+    count_nombre_etablissements_ouverts << count_nombre_etablissements\
+    << create_etablissement_table
+    add_liste_adresses << add_liste_enseignes << count_nombre_etablissements_ouverts
+    create_elastic_index << create_siege_only_table << add_liste_adresses
+    check_elastic_index << fill_elastic_index << create_elastic_index
+    execute_aio_container << update_color_file << check_elastic_index
+    send_email << execute_aio_container
```
Comment (geoffreyaldebert) [inline DAG-insert-elk-sirene.py:159]: It's ok but I prefer [this kind of syntax](https://github.com/etalab/dag_inpi_data/blob/56c4264399f3f0f79e9dc6e5f1d562d6f3809640/DAG.py#L312) which is more clear when you have many tasks

File: data_enrichment.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    with open(f"{labels_file_path}{file_name}") as json_file:
+        file_decoded = json.load(json_file)
+    return file_decoded
+
+
+sections_NAF = load_file("sections_codes_naf.json")
+
+
+# Nom complet
+def create_nom_complet(nature_juridique_unite_legale=None, nom=None, nom_usage=None,
```
Comment (geoffreyaldebert) [inline data_enrichment.py:16]: when you do not have `sigle` or `formatted_sigle` you have something like `Pierre Dupont ()`. It's not a very big problem but not clean

File: utils.py [hunk 3]
```diff
@@ -39,103 +51,643 @@ def get_colors(**kwargs):
...
+                    ''')
+    siren_db_cursor.execute(
+        '''INSERT INTO count_etab_ouvert (siren, count) SELECT siren, count(*) as count FROM siret WHERE etat_administratif_etablissement = 'A' GROUP BY siren;''')
+    commit_and_close_conn(siren_db_conn)
+
+
+def add_liste_enseignes():
+    siren_db_conn, siren_db_cursor = connect_to_db()
+    # Add liste d'enseignes for each établissement
+    add_enseigne = f'''ALTER TABLE siret ADD COLUMN enseignes GENERATED ALWAYS AS
```
Comment (geoffreyaldebert) [inline utils.py:392]: I think this ALTER TABLE is not necessary. We can resolve the creation of the `enseignes` fields at the querying just before indexing into ES.

Comment (geoffreyaldebert) [reply utils.py:392]: Same approach that we use for dirigeants with json objects and subqueries

File: utils.py [hunk 4]
```diff
@@ -39,103 +51,643 @@ def get_colors(**kwargs):
...
+                                                SELECT add_adresse_complete
+                                                (COALESCE(complement_adresse,''),COALESCE(numero_voie,''),
+                                                COALESCE(indice_repetition,''), COALESCE(type_voie,''),
+                                                COALESCE(libelle_voie,''), COALESCE(libelle_commune,''),
+                                                COALESCE(libelle_cedex,''), COALESCE(distribution_speciale,''),
+                                                COALESCE(commune,''), COALESCE(cedex,''), COALESCE(libelle_commune_etranger,''),
+                                                COALESCE(libelle_pays_etranger,'')))''')
+    # Create adresses table with grouped addresses for each siren
+    siren_db_cursor.execute(f'''DROP TABLE IF EXISTS adresses''')
+    siren_db_cursor.execute(f'''CREATE TABLE adresses (siren VARCHAR(10),
```
Comment (geoffreyaldebert) [inline utils.py:426]: same here. Not necessary

File: utils.py [hunk 5]
```diff
@@ -39,103 +51,643 @@ def get_colors(**kwargs):
...
+        elastic_bulk_size=ELASTIC_BULK_SIZE,
+    )
+    create_index.execute()
+
+
+def fill_elastic_index(**kwargs):
+    next_color = kwargs["ti"].xcom_pull(key="next_color", task_ids="get_colors")
+    elastic_index = "siren-" + next_color
+    siren_db_conn, siren_db_cursor = connect_to_db()
+    siren_db_cursor.execute(f'''
```
Comment (geoffreyaldebert) [inline utils.py:611]: I like SQL 😻

File: utils.py [hunk 6]
```diff
@@ -39,103 +51,643 @@ def get_colors(**kwargs):
...
+            ul.categorie_entreprise as categorie_entreprise,
+            ul.etat_administratif_unite_legale as etat_administratif_unite_legale,
+            ul.nom_raison_sociale as nom_raison_sociale,
+            ul.nature_juridique_unite_legale as nature_juridique_unite_legale,
+            ul.activite_principale_unite_legale as activite_principale_unite_legale,
+            ul.economie_sociale_solidaire_unite_legale as 
+            economie_sociale_solidaire_unite_legale,
+            (SELECT count FROM count_etab ce WHERE ce.siren = st.siren) as nombre_etablissements,
+            (SELECT count FROM count_etab_ouvert ceo WHERE ceo.siren = st.siren) as nombre_etablissements_ouverts,
+            (SELECT liste_enseignes FROM enseignes le WHERE le.siren = st.siren) as liste_enseignes,
```
Comment (geoffreyaldebert) [inline utils.py:651]: something like (not tested) : 
```
(SELECT json_group_array(
                json_object(
                    \'enseigne_1\', enseigne_1,
                    \'enseigne_2\', enseigne_2,
                    \'enseigne_3\', enseigne_3
                    )
                ) FROM 
                (
                    SELECT enseigne_1, enseigne_2, enseigne_3 from siret
                    WHERE siren = st.siren
                )
            ) as liste_enseignes
```

File: utils.py [hunk 7]
```diff
@@ -39,103 +51,643 @@ def get_colors(**kwargs):
...
+            (SELECT liste_adresses FROM adresses la WHERE la.siren = st.siren) as liste_adresses,
```
Comment (geoffreyaldebert) [inline utils.py:652]: something like (not tested) : 
```
(SELECT json_group_array(
                json_object(
                    \'indice_repetition\', indice_repetition,
                    \'libelle_voie\', libelle_voie,
                    \'libelle_cedex\', libelle_cedex,
                    ...
                    )
                ) FROM 
                (
                    SELECT indice_repetition, libelle_voie, libelle_cedex, ... from siret
                    WHERE siren = st.siren
                )
            ) as liste_adresses
```

Comment (geoffreyaldebert) [reply utils.py:652]: then you need to process it in a python function

File: utils.py (hunk 6 above)
Comment (geoffreyaldebert) [reply utils.py:651]: then you need to process it in a python function

**Review by geoffreyaldebert — APPROVED** [review]

File: DAG-insert-elk-sirene.py (hunk 1 above)
Comment (HAEKADI) [reply DAG-insert-elk-sirene.py:159]: Oops, I already updated it in the version I didn't push 😅

File: data_enrichment.py (hunk 2 above)
Comment (HAEKADI) [reply data_enrichment.py:16]: Yeah ! Nice catch, I added a `string.replace()`.

File: utils.py (hunk 3 above)
Comment (HAEKADI) [reply utils.py]: Tested and implemented!

File: utils.py (hunk 6 above)
Comment (HAEKADI) [reply utils.py]: Yes, I used the same logic!

**Comment by HAEKADI** [comment]
Testing the search algo with the new and improved index yielded the following results which seem satisfactory.
Screenshot 2022-08-17 at 17 21 19

To make sure the new indices and the preprocessing workflow do not degrade data quality, a thorough comparaison between a dev index and the current production index has been made (Jupiter notebook link).

Siren numbers were chosen at random from the `unite_legale stock`, then the corresponding indexed data in both the dev index and prod index have been compared, mainly focusing on `nom_complet` and `adresse_etablissment` fields.

It seems that the new index is more reliable and up to date than the production status quo.

The next step is to test the new workflow in staging to make sure it is bullet proof.

File: data_enrichment.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    if nom_raison_sociale:
+        name = nom_raison_sociale
+
+    if sigle:
+        name = f"{name} ({sigle})"
+    return name.lower().strip() if name else name
+
+
+# Entrepreneur individuel
+def create_entrepreneur_individuel(nature_juridique_unite_legale):
```
Comment (XavierJp) [inline data_enrichment.py:42]: `create`dans le nom est un peu surprenant 

Pourquoi pas `is_` ou `est_`

File: data_enrichment.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    if nature_juridique_unite_legale in ["1", "10", "1000"]:
+        return "true"
+    else:
+        return "false"
+
+
+# Section activité principale
+def create_section(activite_principale_unite_legale):
```
Comment (XavierJp) [inline data_enrichment.py:50]: En fin de compte avoir `create_` partout n'apporte pas grand chose

On pourrait imaginer `is_` / `est_` pour des booleans des `label_from` pour de la conversion de code en string, ou des `format_`pour du formatage de string

File: elasticsearch/create_siren.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        if not self.elastic_url:
+            raise ValueError("Please provide elasticsearch url endpoint")
+
+        # if self.elastic_index_shards is not None:
+        if Index(self.elastic_index).exists():
+            logging.info(f"Index  {self.elastic_index} already exists! Deleting...")
+            Index(self.elastic_index).delete()
+            logging.info(f"Index {self.elastic_index} deleted!")
+        logging.info(f"Creating {self.elastic_index} index!")
+        Siren.init()
```
Comment (XavierJp) [inline elasticsearch/create_siren.py:74]: Ce `Siren.init()`est bien mystérieux, on pourrait choisir un nom décrivant mieux l'action

File: elasticsearch/index_doc.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+            include_meta=True
+        )
+
+
+def index_by_chunk(cursor, elastic_connection, elastic_bulk_size, elastic_index):
+    i = 0
+    res = 1
+    while res:
+        res = cursor.fetchmany(elastic_bulk_size)
+        columns = tuple([x[0] for x in cursor.description])
```
Comment (XavierJp) [inline elasticsearch/index_doc.py:21]: Same, too many unclear variables names : index_by_chunk, doc_generator, process_doc, res, res2

File: elasticsearch/mapping_siren.py [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        french_elision,
+        french_stop,
+        "icu_folding",
+        french_synonym,
+        "asciifolding",
+        french_stemmer,
+    ],
+)
+
+
```
Comment (XavierJp) [inline elasticsearch/mapping_siren.py:58]: Really cool to have a model for a document 👍

File: elasticsearch/process_doc.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    create_adresse_complete,
+    create_coordonnees,
+    create_departement,
+    create_entrepreneur_individuel,
+    create_nom_complet,
+    create_section,
+)
+from dag_datalake_sirene.helpers.single_dispatch_funcs import get_string
+
+
```
Comment (XavierJp) [inline elasticsearch/process_doc.py:13]: Same here, too many unclear variables

**Review by XavierJp — CHANGES_REQUESTED** [review]

File: elasticsearch/create_siren.py (hunk 10 above)
Comment (HAEKADI) [reply elasticsearch/create_siren.py:74]: Comme quoi par exemple ? Le Siren est le nom de la classe Elastic

**Comment by HAEKADI** [comment]
@XavierJp I applied the changes as discussed.
In the meantime, I started the workflow and will do another round of testing just to make sure everything is a-okay.

**Review by XavierJp — APPROVED** [review]

**Comment by HAEKADI** [comment]
Tests performed the same after refactoring. Merging 🚀

**Pull request merged**